### PR TITLE
Fixes Keras import issue (8373)

### LINKED ISSUE #8373: Unable to import keras functional model
I am trying to import a keras functional model DEEPFM  based on [paper](https://arxiv.org/pdf/1703.04247.pdf) into DL4J. There are 3 inputs, 1 numerical and 2 embedding inputs. The model has 3 parts :
1) Linear part : 
2) Factorization Machine
3) Deep NN Part
Attaching the keras model summary and JSON file for reference:

```
Model: "model_8"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
g (InputLayer)                  (None, 1)            0                                            
__________________________________________________________________________________________________
campaign_id (InputLayer)        (None, 1)            0                                            
__________________________________________________________________________________________________
client_id (InputLayer)          (None, 1)            0                                            
__________________________________________________________________________________________________
g_2d (Dense)                    (None, 8)            16          g[0][0]                          
__________________________________________________________________________________________________
reshape_6 (Reshape)             (None, 1, 8)         0           g_2d[0][0]                       
__________________________________________________________________________________________________
campaign_idemb_2d (Embedding)   (None, 1, 8)         1968        campaign_id[0][0]                
__________________________________________________________________________________________________
client_idemb_2d (Embedding)     (None, 1, 8)         31889352    client_id[0][0]                  
__________________________________________________________________________________________________
concat_embed_2d (Concatenate)   (None, 3, 8)         0           reshape_6[0][0]                  
                                                                 campaign_idemb_2d[0][0]          
                                                                 client_idemb_2d[0][0]            
__________________________________________________________________________________________________
flat_embed_2d (Flatten)         (None, 24)           0           concat_embed_2d[0][0]            
__________________________________________________________________________________________________
dropout_3 (Dropout)             (None, 24)           0           flat_embed_2d[0][0]              
__________________________________________________________________________________________________
sum_of_tensors (Lambda)         multiple             0           concat_embed_2d[0][0]            
                                                                 square_of_tensors[0][0]          
                                                                 lambda_2[0][0]                   
__________________________________________________________________________________________________
square_of_tensors (Lambda)      (None, 3, 8)         0           concat_embed_2d[0][0]            
__________________________________________________________________________________________________
dense_3 (Dense)                 (None, 128)          3200        dropout_3[0][0]                  
__________________________________________________________________________________________________
multiply_6 (Multiply)           (None, 8)            0           sum_of_tensors[0][0]             
                                                                 sum_of_tensors[0][0]             
__________________________________________________________________________________________________
batch_normalization_3 (BatchNor (None, 128)          512         dense_3[0][0]                    
__________________________________________________________________________________________________
subtract_2 (Subtract)           (None, 8)            0           multiply_6[0][0]                 
                                                                 sum_of_tensors[1][0]             
__________________________________________________________________________________________________
dropout_4 (Dropout)             (None, 128)          0           batch_normalization_3[0][0]      
__________________________________________________________________________________________________
campaign_idemb_linear (Embeddin (None, 1, 1)         246         campaign_id[0][0]                
__________________________________________________________________________________________________
client_idemb_linear (Embedding) (None, 1, 1)         3986169     client_id[0][0]                  
__________________________________________________________________________________________________
lambda_2 (Lambda)               (None, 8)            0           subtract_2[0][0]                 
__________________________________________________________________________________________________
dense_4 (Dense)                 (None, 128)          16512       dropout_4[0][0]                  
__________________________________________________________________________________________________
g_linear (Dense)                (None, 1)            2           g[0][0]                          
__________________________________________________________________________________________________
reshape_4 (Reshape)             (None, 1)            0           campaign_idemb_linear[0][0]      
__________________________________________________________________________________________________
reshape_5 (Reshape)             (None, 1)            0           client_idemb_linear[0][0]        
__________________________________________________________________________________________________
batch_normalization_4 (BatchNor (None, 128)          512         dense_4[0][0]                    
__________________________________________________________________________________________________
linear_output (Add)             (None, 1)            0           g_linear[0][0]                   
                                                                 reshape_4[0][0]                  
                                                                 reshape_5[0][0]                  
__________________________________________________________________________________________________
fm_2d_output (Reshape)          (None, 1)            0           sum_of_tensors[2][0]             
__________________________________________________________________________________________________
deep_output (Dense)             (None, 1)            129         batch_normalization_4[0][0]      
__________________________________________________________________________________________________
3_combined (Concatenate)        (None, 3)            0           linear_output[0][0]              
                                                                 fm_2d_output[0][0]               
                                                                 deep_output[0][0]                
__________________________________________________________________________________________________
deepfm_output (Dense)           (None, 1)            4           3_combined[0][0]                 
==================================================================================================
Total params: 35,898,622
Trainable params: 35,898,110
Non-trainable params: 512
```

Attaching the scala code. I defined the **sum of tensors, square of tensors and lambda_2** layer which were then registered in scala while importing the model in DL4J. 
```
class tensors_sum extends SameDiffLambdaLayer {
  override def defineLayer(sd: SameDiff, x: SDVariable): SDVariable = x.sum(false,1)
  override def getOutputType(layerIndex: Int, inputType: InputType): InputType = inputType
}


class tensors_square extends SameDiffLambdaLayer {
  override def defineLayer(sd: SameDiff, x: SDVariable): SDVariable = x mul x
  override def getOutputType(layerIndex: Int, inputType: InputType): InputType = inputType
}

class lambda1 extends SameDiffLambdaLayer {
  override def defineLayer(sd: SameDiff, x: SDVariable): SDVariable = x.mul(0.5)

  override def getOutputType(layerIndex: Int, inputType: InputType): InputType = inputType
}

KerasLayer.registerLambdaLayer("sum_of_tensors", new tensors_sum())
    KerasLayer.registerLambdaLayer("square_of_tensors", new tensors_square())
    KerasLayer.registerLambdaLayer("lambda_2", new lambda1())
    val resourcesPath = getClass.getResource("/deep_fm_sample.h5").getPath
val model1 = KerasModelImport.importKerasModelAndWeights(resourcesPath,false)


val gu1=Nd4j.zeros(1)
    val guin = gu1.reshape(1,1)


     val c1=Nd4j.zeros(1)
    val c1in=c1.reshape(1,1)

    val c2=Nd4j.zeros(1)
    val c2in=c2.reshape(1,1)


  
    model1.outputSingle(guin,c1in,c2in)
```
 After executing this I got the following error:
```
Error at [/home/jenkins/agent/workspace/ing4j_deeplearning4j-1.0.0-beta5/libnd4j/include/ops/declarable/generic/transforms/concat.cpp:155:0]:
CONCAT op: all input arrays must have the same dimensions (except those on input axis) !
Exception in thread "main" org.nd4j.linalg.exception.ND4JIllegalStateException: Op name concat failed to execute. You can't execute non-inplace CustomOp without outputs being specified
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:1679)
	at org.nd4j.linalg.factory.Nd4j.exec(Nd4j.java:6606)
	at org.nd4j.linalg.cpu.nativecpu.CpuNDArrayFactory.concat(CpuNDArrayFactory.java:557)
	at org.nd4j.linalg.factory.Nd4j.concat(Nd4j.java:4917)
	at org.nd4j.linalg.factory.BaseNDArrayFactory.hstack(BaseNDArrayFactory.java:963)
	at org.nd4j.linalg.factory.Nd4j.hstack(Nd4j.java:4673)
	at org.deeplearning4j.nn.graph.vertex.impl.MergeVertex.doForward(MergeVertex.java:117)
	at org.deeplearning4j.nn.graph.ComputationGraph.outputOfLayersDetached(ComputationGraph.java:2379)
	at org.deeplearning4j.nn.graph.ComputationGraph.output(ComputationGraph.java:1792)
	at org.deeplearning4j.nn.graph.ComputationGraph.outputSingle(ComputationGraph.java:1774)
	at org.deeplearning4j.nn.graph.ComputationGraph.outputSingle(ComputationGraph.java:1760)
	at org.deeplearning4j.nn.graph.ComputationGraph.outputSingle(ComputationGraph.java:1638)
	at Demo1$.main(Demo1.scala:76)
	at Demo1.main(Demo1.scala)
```

[model.zip](https://github.com/eclipse/deeplearning4j/files/3824344/model.zip)

I might be doing something wrong here if its not the import issue. Please help

### PR DESCRIPTION
**Pull request by farizrahman4u**
Fixes [8373](https://github.com/eclipse/deeplearning4j/issues/8373)

### CODE REVIEW

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/graph/ComputationGraph.java [hunk 1]
```diff
@@ -795,6 +795,9 @@ public void initGradientsView() {
     }
 
     protected int[] getOutputLayerIndices(){
+        for (String s: verticesMap.keySet()){
```
Comment (agibsonccc) [inline deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/graph/ComputationGraph.java]: Get rid of debug statements

**Review by agibsonccc — CHANGES_REQUESTED** [review]

**Comment by agibsonccc** [comment]
@maxpumperla pending the model being added to test resources anything else you want done here?

**Comment by maxpumperla** [comment]
@agibsonccc just an actual test. it doesn't have to be the one from the issue, a minimal keras example showing the intended behavior suffices. we've had the "will add test resources later" (but never did) situation a bit too often for my taste. it's normal to forget such things, so let's just add a quick test now.

**Comment by agibsonccc** [comment]
@maxpumperla cc @AlexDBlack  is this on our internal tools anywhere yet?  @sshepel how is this PR looking on jenkins?

**Review by maxpumperla — APPROVED** [review]
thanks for the test

**Comment by AlexDBlack** [comment]
I'll merge this. Waiting for CI would be better, but it's probably not reliable enough to get all the way to keras import tests without hitting an issue.
Any issues should be picked up in the pre-release QA we're doing now anyway.

**Pull request merged**